### PR TITLE
[leveldb] Install util header files

### LINKED ISSUE #18493: Leveldb is installed, but the library is missing
The leveldb library is installed, but only include/leveldb is included, there is no util 
How should I add the util library? 
2~DEJ%C26 OZ)_$2{}0NNP5

### PR DESCRIPTION
**Pull request by JonLiu1993**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #18493

The leveldb library is installed, but only include/leveldb is included, there is no util, I install util header files in _DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/leveldb_  directroy and Modernize portfile.cmake

### CODE REVIEW

File: ports/leveldb/fix-util-install.patch [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+       "${PROJECT_SOURCE_DIR}/${LEVELDB_PUBLIC_INCLUDE_DIR}/table.h"
+       "${PROJECT_SOURCE_DIR}/${LEVELDB_PUBLIC_INCLUDE_DIR}/write_batch.h"
++      "${PROJECT_SOURCE_DIR}/util/arena.h"
++      "${PROJECT_SOURCE_DIR}/util/coding.h"
++      "${PROJECT_SOURCE_DIR}/util/crc32c.h"
++      "${PROJECT_SOURCE_DIR}/util/hash.h"
++      "${PROJECT_SOURCE_DIR}/util/logging.h"
++      "${PROJECT_SOURCE_DIR}/util/mutexlock.h"
++      "${PROJECT_SOURCE_DIR}/util/no_destructor.h"
++      "${PROJECT_SOURCE_DIR}/util/random.h"
```
Comment (JackBoosY) [inline ports/leveldb/fix-util-install.patch:16]: This seems incorrect, all the added headers should be installed into `include/leveldb/util`.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/leveldb/fix-util-install.patch (hunk 1 above)
Comment (JonLiu1993) [reply ports/leveldb/fix-util-install.patch]: Done

Comment (JackBoosY) [reply ports/leveldb/fix-util-install.patch]: It's not done:
```
leveldb:x86-windows:leveldb/arena.h
leveldb:x86-windows:leveldb/c.h
leveldb:x86-windows:leveldb/cache.h
leveldb:x86-windows:leveldb/coding.h
leveldb:x86-windows:leveldb/comparator.h
leveldb:x86-windows:leveldb/crc32c.h
leveldb:x86-windows:leveldb/db.h
leveldb:x86-windows:leveldb/dumpfile.h
leveldb:x86-windows:leveldb/env.h
leveldb:x86-windows:leveldb/env_posix_test_helper.h
leveldb:x86-windows:leveldb/env_windows_test_helper.h
leveldb:x86-windows:leveldb/export.h
leveldb:x86-windows:leveldb/filter_policy.h
leveldb:x86-windows:leveldb/hash.h
leveldb:x86-windows:leveldb/histogram.h
leveldb:x86-windows:leveldb/iterator.h
leveldb:x86-windows:leveldb/logging.h
leveldb:x86-windows:leveldb/mutexlock.h
leveldb:x86-windows:leveldb/no_destructor.h
leveldb:x86-windows:leveldb/options.h
leveldb:x86-windows:leveldb/posix_logger.h
leveldb:x86-windows:leveldb/random.h
leveldb:x86-windows:leveldb/slice.h
leveldb:x86-windows:leveldb/status.h
leveldb:x86-windows:leveldb/table.h
leveldb:x86-windows:leveldb/table_builder.h
leveldb:x86-windows:leveldb/testharness.h
leveldb:x86-windows:leveldb/testutil.h
leveldb:x86-windows:leveldb/windows_logger.h
leveldb:x86-windows:leveldb/write_batch.h
```

Comment (JackBoosY) [reply ports/leveldb/fix-util-install.patch]: I don't think you solved my question:
```
if(LEVELDB_INSTALL)
       "${PROJECT_SOURCE_DIR}/${LEVELDB_PUBLIC_INCLUDE_DIR}/table_builder.h"
       "${PROJECT_SOURCE_DIR}/${LEVELDB_PUBLIC_INCLUDE_DIR}/table.h"
       "${PROJECT_SOURCE_DIR}/${LEVELDB_PUBLIC_INCLUDE_DIR}/write_batch.h"
+      "${PROJECT_SOURCE_DIR}/util/arena.h"
+      "${PROJECT_SOURCE_DIR}/util/coding.h"
+      "${PROJECT_SOURCE_DIR}/util/crc32c.h"
+      "${PROJECT_SOURCE_DIR}/util/hash.h"
+      "${PROJECT_SOURCE_DIR}/util/logging.h"
+      "${PROJECT_SOURCE_DIR}/util/mutexlock.h"
+      "${PROJECT_SOURCE_DIR}/util/no_destructor.h"
+      "${PROJECT_SOURCE_DIR}/util/random.h"
+      "${PROJECT_SOURCE_DIR}/util/env_posix_test_helper.h"
+      "${PROJECT_SOURCE_DIR}/util/env_windows_test_helper.h"
+      "${PROJECT_SOURCE_DIR}/util/histogram.h"
+      "${PROJECT_SOURCE_DIR}/util/posix_logger.h"
+      "${PROJECT_SOURCE_DIR}/util/testharness.h"
+      "${PROJECT_SOURCE_DIR}/util/testutil.h"
+      "${PROJECT_SOURCE_DIR}/util/windows_logger.h"
+      "${PROJECT_SOURCE_DIR}/table/block.h"
+      "${PROJECT_SOURCE_DIR}/table/block_builder.h"
+      "${PROJECT_SOURCE_DIR}/table/filter_block.h"
+      "${PROJECT_SOURCE_DIR}/table/format.h"
+      "${PROJECT_SOURCE_DIR}/table/iterator_wrapper.h"
+      "${PROJECT_SOURCE_DIR}/table/merger.h"
+      "${PROJECT_SOURCE_DIR}/table/two_level_iterator.h"
     DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/leveldb
   )
```
This behavor will install all the headers into _include/leveldb_ **without** a sub folder.

File: ports/leveldb/fix-util-install.patch [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
++      "${PROJECT_SOURCE_DIR}/util/env_posix_test_helper.h"
++      "${PROJECT_SOURCE_DIR}/util/env_windows_test_helper.h"
++      "${PROJECT_SOURCE_DIR}/util/histogram.h"
++      "${PROJECT_SOURCE_DIR}/util/posix_logger.h"
++      "${PROJECT_SOURCE_DIR}/util/testharness.h"
++      "${PROJECT_SOURCE_DIR}/util/testutil.h"
++      "${PROJECT_SOURCE_DIR}/util/windows_logger.h"
++    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/util
```
Comment (JackBoosY) [inline ports/leveldb/fix-util-install.patch:27]:
```suggestion
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/leveldb/util
```

File: ports/leveldb/fix-util-install.patch [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
++  install(
++    FILES
++      "${PROJECT_SOURCE_DIR}/table/block.h"
++      "${PROJECT_SOURCE_DIR}/table/block_builder.h"
++      "${PROJECT_SOURCE_DIR}/table/filter_block.h"
++      "${PROJECT_SOURCE_DIR}/table/format.h"
++      "${PROJECT_SOURCE_DIR}/table/iterator_wrapper.h"
++      "${PROJECT_SOURCE_DIR}/table/merger.h"
++      "${PROJECT_SOURCE_DIR}/table/two_level_iterator.h"
++    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/table
```
Comment (JackBoosY) [inline ports/leveldb/fix-util-install.patch:39]:
```suggestion
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/leveldb/table
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — DISMISSED** [review]

**Comment by ras0219-msft** [comment]
It looks like these header files are intended to be private (`https://github.com/google/leveldb/blob/main/CMakeLists.txt`). Can we open an issue upstream and confirm whether they want these util headers to be available for consumers or not?

**Comment by JonLiu1993** [comment]
> It looks like these header files are intended to be private (`https://github.com/google/leveldb/blob/main/CMakeLists.txt`)

@ras0219-msft, Thanks for the review, I've filed an [issue ](https://github.com/google/leveldb/issues/983)upstream to confirm

**Comment by JackBoosY** [comment]
Draft this PR until the upstream has response.

**Comment by JackBoosY** [comment]
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.

**Comment by JonLiu1993** [comment]
Regarding my changes, I have submitted an [issue](https://github.com/google/leveldb/issues/983) upstream to confirm whether it is feasible, but there is still no response after two years, so I reopened the PR.

**Review by Cheney-W — APPROVED** [review]

**Pull request merged**